### PR TITLE
Add support for tests

### PR DESCRIPTION
**Pull request by kant2002**
This is minimum amount of changes which require to 
- add ability to have in-memory drawing device
- Be able to load data from AGG files multiple times in the same application. 

I test these changes on simple tests which just draw different luck/morale modifiers depending on the hero stats.
These test definitely do not optimal, and maybe useful, but serves me as PoC for whole work.

### CODE REVIEW

File: src/fheroes2/agg/agg.cpp [hunk 1]
```diff
@@ -803,6 +804,7 @@ void AGG::Quit( void )
     wav_cache.clear();
     mid_cache.clear();
     loop_sounds.clear();
+    fheroes2::AGG::clearICNSprites();
```
Comment (ihhub) [inline src/fheroes2/agg/agg.cpp:807]: Let's call it from fheroes2.cpp. We should avoid dependencies.

File: src/fheroes2/agg/agg_image.cpp [hunk 2]
```diff
@@ -878,4 +878,9 @@ namespace fheroes2
             return height;
         }
     }
+
+    void AGG::clearICNSprites()
+    {
+        _icnVsSprite.clear();
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:884]: We should clean TIL images as well. In this case we can call this function as `clear`. Also please put it under the namespace above.

File: src/engine/screen.h [hunk 3]
```diff
@@ -124,7 +124,16 @@ namespace fheroes2
 
         friend BaseRenderEngine & engine();
         friend Cursor & cursor();
-
+#if CAN_OVERRIDE_RENDERING
```
Comment (ihhub) [inline src/engine/screen.h:127]: Could we please avoid using macros her? Could you please explain the intention of this change?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @kant2002 , I left multiple comments here. Would you mind to take a look once you have time?

File: src/engine/screen.h (hunk 3 above)
Comment (kant2002) [reply src/engine/screen.h:127]: I have to provide custom display (renderer) and custom cursor for tests. Currently there no way to provide different display and cursor without changing how these objects constructed. There other way where I change existing way how to construct rendering pipeline, but I decide that this is less invasive way. These code needed only for tests, so I decide to not include it in production version

File: src/fheroes2/agg/agg_image.cpp (hunk 2 above)
Comment (kant2002) [reply src/fheroes2/agg/agg_image.cpp]: This seems to be redundant right now.

File: src/engine/screen.h (hunk 3 above)
Comment (kant2002) [reply src/engine/screen.h:127]: Look at the https://github.com/kant2002/fheroes2/blob/kant/example-tests/tests/fheroes2.tests/test.cpp#L67 how I currently manage to create tests

Comment (ihhub) [reply src/engine/screen.h:127]: Let me please come up with a solution to replace a singleton or to support multiple instances of Display. I'm sure we could do this in non-hacky way :)

**Comment by kant2002** [comment]
From Discord discussion
- @ihhub : actually we need to replace renderEngine instead of Display as display is just an image

**Comment by ihhub** [comment]
I'm closing this pull request as we merge similar code for the same problem.

**Pull request closed without merging**